machine_learning/automatic_differentiation.py

```diff
@@ -0,0 +1,322 @@
+"""
+Demonstration of the Automatic Differentiation (Reverse mode).
+
+Reference: https://en.wikipedia.org/wiki/Automatic_differentiation
+
+Author: Poojan Smart
+Email: smrtpoojan@gmail.com
+
+Examples:
+
+>>> with GradientTracker() as tracker:
+...     a = Variable([2.0, 5.0])
+...     b = Variable([1.0, 2.0])
+...     m = Variable([1.0, 2.0])
+...     c = a + b
+...     d = a * b
+...     e = c / d
+>>> print(tracker.gradient(e, a))
+[-0.25 -0.04]
+>>> print(tracker.gradient(e, b))
+[-1.   -0.25]
+>>> print(tracker.gradient(e, m))
+None
+
+>>> with GradientTracker() as tracker:
+...     a = Variable([[2.0, 5.0]])
+...     b = Variable([[1.0], [2.0]])
+...     c = a @ b
+>>> print(tracker.gradient(c, a))
+[[1. 2.]]
+>>> print(tracker.gradient(c, b))
+[[2.]
+ [5.]]
+
+>>> with GradientTracker() as tracker:
+...     a = Variable([[2.0, 5.0]])
+...     b = a ** 3
+>>> print(tracker.gradient(b, a))
+[[12. 75.]]
+"""
+from __future__ import annotations
+
+from enum import Enum
+from types import TracebackType
+from typing import Self
```

Comment:
Ruff was upgraded for Py3.12 but mypy has not yet been upgraded.
```suggestion
from typing_extensions import Self  # noqa: UP035
```